### ISSUE TITLE
isFilePasswordProtectedAtPath: returns NO for AES password-protected zip files

### ISSUE THREAD

**Issue by hacknicity** [ZipArchive/ZipArchive#445]
I'm a new user of this library (thanks *so* much for it!) and am using it to create an archive that stores several data files from my app. I want the user to be able to optionally password-protect the backup. Thus, when I restore from my backup I want to test whether I need to ask them for a password. In my testing, `isFilePasswordProtectedAtPath:` method always returns `NO`, whether the archive was created with a password or not. 

Looking at some of the history of the GitHub issues in this project it sounds like a change was made at some point to switch to AES compression. I'm creating my archive with `createZipFileAtPath:withFilesAtPaths:` so don't have a way to opt out of AES encryption.

The `testIsFilePasswordProtectedAtPath` unit test uses `TestPasswordArchive.zip`, which seems to have been created before AES compression was added.

I've done a little debugging and `isFilePasswordProtectedAtPath:` is failing here:

~~~~
    int ret = unzGoToFirstFile(zip);
    if (ret == UNZ_OK) {
        do {
            ret = unzOpenCurrentFile(zip);
            if (ret != UNZ_OK) {
                return NO;                    // <<<<<<  returns here for AES files
            }
~~~~

`unzOpenCurrentFile` calls through to `unzOpenCurrentFile3` passing a nil `password` parameter. And `unzOpenCurrentFile3` has this code which gives up since there is no password parameter:

~~~~
    compression_method = s->cur_file_info.compression_method;
#ifdef HAVE_AES
    if (compression_method == AES_METHOD)
    {
        compression_method = s->cur_file_info_internal.aes_compression_method;
        if (password == NULL)
        {
            return UNZ_PARAMERROR;         // <<<<<<< returns HERE
        }
    }
#endif
~~~~

It looks like this check was added in f13ff57 when the unzip library was updated.

**Comment by jhudsonWA** [comment]
I have a branch to switch to version 2.0 of the upstream library we use (minizip) to do the compression/decompression. I'll add tests to check to see if this is fixed.

**Comment by hacknicity** [comment]
Thank you. It feels unlikely, given the nature of the issue, but you never know!

For what it's worth, simply commenting out the NULL check has allowed me to reliably detect AES protected archives. I'm not sure if that will actually break anything (I've not go as far as unzipping my files yet!) Of course, that is not really a solution that could be used since it affects the upstream library.

**Comment by hacknicity** [comment]
It seems that the minizip library doesn't really support opening AES protected files without providing the password.

A quick and naive look at the code makes me wonder if there may be a solution by changing `isFilePasswordProtectedAtPath:` to use `unzOpenCurrentFilePassword` instead of `unzOpenCurrentFile` and provide a made up password. Looking at the code, I think that might result in `unzOpenCurrentFile3` failing with `UNZ_BADPASSWORD` (for the AES case).

But I suspect that would break non-AES passwords where `unzOpenCurrentFile3` uses the `password` parameter:

~~~~
            s->pcrc_32_tab = (const z_crc_t*)get_crc_table();
            init_keys(password, s->keys, s->pcrc_32_tab);
~~~~

**Comment by hacknicity** [comment]
It would need more thorough testing, but changing `isFilePasswordProtectedAtPath:`'s

~~~~
            ret = unzOpenCurrentFile(zip);
~~~~

to 

~~~~
            ret = unzOpenCurrentFilePassword(zip, "dummypassword");
            if (ret == UNZ_BADPASSWORD) {
                return YES;
            } else if (ret != UNZ_OK) {
                return NO;
            }
~~~~

does *seem* to work. I.e. I think it correctly identifies non-password-protected files, AES-protected files and the non-AES-protected file `TestPasswordArchive.zip` from the unit tests. I'm not actually running the unit tests, I'm just hacking about in my own app!

The code at the end of `unzOpenCurrentFile3` that uses the dummy password seems to be benign when used from `isFilePasswordProtectedAtPath:` (presumably since just testing for the password doesn't try and unarchive anything).

**Comment by hacknicity** [comment]
My final comment before calling it a day and giving you a chance to do your own testing: my suggested workaround even seems to work if the archive was protected with an AES password of `dummypassword`! An empty password also seems to work and is probably a better choice:

~~~~
            ret = unzOpenCurrentFilePassword(zip, "");
~~~~

**Comment by Coeur** [comment]
Hi @hacknicity, and thank you for your interest in ZipArchive. Sorry for replying so late.

Regarding the blame, the check was added here: https://github.com/nmoinvaz/minizip/commit/d6aed934e6f16fbe0b72af622530904ec0a7ce1d
Which refers to this issue: https://github.com/nmoinvaz/minizip/issues/76

Now, regarding if `isFilePasswordProtectedAtPath` is working or not for AES encrypted files, that's a good remark. I haven't look at it yet, but it's worth investigating. I'll look at it tonight.

**Comment by Coeur** [comment]
Fixed in 2.1.2: 1584da8c39006555ba5e7e7ea6e3c5e9ed39c2d7
